Groups and Aliases Specification Updates (#298)
design-tokens/community-group: technical-reports/format/groups.md

# Groups

Groups organize design tokens into logical collections and provide a hierarchical structure for token files. Groups are arbitrary and tools SHOULD NOT use them to infer the type or purpose of design tokens.

## Group Structure

A group is identified as a JSON object that does NOT contain a [`$value`](design-token#name-and-value) property. Groups MAY contain:

- **Child tokens** - Objects with a `$value` property
- **Nested groups** - Objects without a `$value` property
- **Group properties** - Properties prefixed with `$` (e.g., `$description`, `$type`)

**Important:** The presence of a `$value` property definitively identifies an object as a token. If an object contains both `$value` and child tokens/groups, this creates an invalid structure where the object cannot be both a token and a group simultaneously. Tools MUST report this as an error.

## Root Tokens in Groups

Groups MAY contain a **root token** alongside child tokens and nested groups. A root token provides a base value for the group while allowing for variants or extensions.

### Syntax for Root Tokens

Groups support root tokens using the reserved name `$root` as the token name:

> <aside class="example">

Author: Sounds good to me! Will update.

<aside class="example">

```json
{
  "spacing": {
    "$description": "All spacing-related design tokens organized by usage context",
    "margin": {
      /* tokens */
    },
    "padding": {
      /* tokens */
    }
  }
}
```

</aside>

#### `$type`

Groups MAY include an optional [`$type`](design-token#type) property that acts as a default type for any tokens within the group that do not explicitly declare their own type. This type inheritance applies to nested groups and their tokens unless overridden.

<aside class="example">

```json
{
  "color": {
    "$type": "color",
    "primary": {
      "$value": {
        "colorSpace": "srgb",
        "components": [0, 0.4, 0.8],
        "hex": "#0066cc"
      }
    },
    "secondary": {
      "$value": {
        "colorSpace": "srgb",
        "components": [0.4, 0.6, 1],
        "hex": "#6699ff"
      }
    },
    "semantic": {
      "success": {
        "$value": {
          "colorSpace": "srgb",
          "components": [0, 0.8, 0.4],
          "hex": "#00cc66"
        }
      },
      "warning": { "$type": "string", "$value": "amber" }
    }
  }
}
```

</aside>

#### `$extends`

> Groups MAY include an optional `$extends` property to inherit tokens and properties from another group. The `$extends` property is syntactic sugar for JSON Schema's `$ref` keyword and follows the same semantic behavior as defined in [[json-schema-2020-12]].

Reviewer: Also, unrelated, but just for clarity specify that it‘s an error if it points to a token. This is strongly suggested by the previous sentence, but just repeating the inverse makes it clear.

```suggestion
Groups MAY include an optional `$extends` property to inherit tokens and properties from another group. `$extends` MUST NOT reference a token. The `$extends` property is syntactic sugar for JSON Schema's `$ref` keyword and follows the same semantic behavior as defined in [[json-schema-2020-12]].
```

Author: Good idea! Clarity is always good.

> Groups MAY include an optional `$extends` property to inherit tokens and properties from another group. The `$extends` property is syntactic sugar for JSON Schema's `$ref` keyword and follows the same semantic behavior as defined in [[json-schema-2020-12]].

Author: I'm struggling to think of a case where you'd want to extend multiple groups; I'm certainly open to using an array here but would rather wait for someone to ask before introducing it.

<aside class="example">

```json
{
  "button": {
    "$type": "color",
    "background": {
      "$value": {
        "colorSpace": "srgb",
        "components": [0, 0.4, 0.8],
        "hex": "#0066cc"
      }
    },
    "text": {
      "$value": {
        "colorSpace": "srgb",
        "components": [1, 1, 1],
        "hex": "#ffffff"
      }
    }
  },
  "button-primary": {
    "$extends": "{button}",
    "background": {
      "$value": {
        "colorSpace": "srgb",
        "components": [0.8, 0, 0.4],
        "hex": "#cc0066"
      }
    }
  }
}
```

> </aside>

Author: Updated to clarify and provide references!

</aside>

In this example:

- `{color.accent.$root}` resolves to `{"colorSpace": "srgb", "components": [0.867, 0, 0], "hex": "#dd0000"}` (the root token)
- `{color.accent.light}` resolves to `{"colorSpace": "srgb", "components": [1, 0.133, 0.133], "hex": "#ff2222"}`
- `{color.accent.dark}` resolves to `{"colorSpace": "srgb", "components": [0.667, 0, 0], "hex": "#aa0000"}`
- `{color.accent}` is an invalid token reference (refers to a group, not a token)

**Rationale:** Using `$root` as a reserved token name eliminates ambiguity between group references and token references while maintaining clear, explicit syntax. The `$` prefix prevents naming conflicts with user-defined tokens and aligns with other reserved properties in the specification.

## Group Properties

> Groups MAY include the following properties:

Author: I think that's a good suggestion!

**Note on Implementation:** While this specification references JSON Schema `$ref` behavior for technical implementation guidance, the user-visible behavior is the straightforward deep merge described above. Tools may implement this merge behavior directly or by leveraging JSON Schema libraries.

##### Inheritance Semantics

Group extension follows **deep merge** behavior where local properties override inherited properties at the same path:

1. **Inheritance:** All tokens and properties from the referenced group are inherited (circular references are not allowed)
2. **Override:** Local tokens and properties replace inherited ones with the same path
3. **Addition:** Local tokens and properties with new paths are added alongside inherited ones

**Override Rules:**

- **Same path = override:** If both the inherited group and local group define a token at the same path, the local definition wins
- **Different paths = merge:** Tokens at different paths coexist in the final result
- **Complete replacement:** When overriding, the entire token definition is replaced (not merged property-by-property)

<aside class="example">

> ```json

Reviewer 2: Should all `aside class=example` have titles for clarity? I've seen recommendations to update some, but not sure if we should apply to all as well.

| -------- | ----------------------------------------------------------------- |
| `true`   | This group is deprecated (no explanation provided)                |
| `string` | This group is deprecated AND this is an explanation               |
| `false`  | This group is NOT deprecated (may override parent group defaults) |

#### `$extensions`

Groups MAY include an optional [`$extensions`](design-token#extensions) property where tools MAY add proprietary, user-, team- or vendor-specific data. Each tool MUST use a vendor-specific key whose value MAY be any valid JSON data.

## Empty Groups

Empty groups (groups containing no tokens or nested groups) are **explicitly permitted**. While they may appear to serve no immediate purpose, they:

- Cause no harm to processing or validation
- Support work-in-progress organization during token development
- May contain metadata via group properties (`$description`, `$extensions`)
- Provide placeholder structure for future token development

<aside class="example">

```json
{
  "experimental": {
    "$description": "Tokens for experimental features - currently empty",
    "$deprecated": "This group is being phased out"
  }
}
```

</aside>

> **Note on Token vs Group Ambiguity:**

Reviewer: ```suggestion
<aside class="note" title="Token vs Group Ambiguity">
```

Objects without a `$value` property are interpreted as groups by definition. This can potentially create ambiguity in cases where a token lacks required properties (such as `$value` or a determinable type) and might be incorrectly parsed as an empty group. Tools SHOULD provide clear error messages when an object appears to be an incomplete token rather than an intentional empty group.

## References and JSON Pointer Integration

The current [token reference syntax](#references-and-json-pointer-integration) using curly braces (`{group.token}`) is maintained for backward compatibility and developer ergonomics. However, tools MAY also support JSON Pointer notation for advanced use cases.

### Current Reference Syntax (Recommended)

> <aside class="example">

Reviewer: Suggestion: I’ll stop commenting `title`s, but here and below, these could be example titles rather than section headings. IMO a section heading should have some meaningful text that stands alone apart from an example. Otherwise, the example should be nested under another heading.

This isn’t W3C styleguide or anything as far as I know. This is just personal preference (which, again, obligatory “I think we’re inconsistent about it”)

```suggestion
<aside class="example" title="Current Reference Syntax (Recommended)">
```

2. **Resolved group's `$type` property** (after extension resolution)
3. **Parent group's `$type` property** (walking up the hierarchy)
4. **Token is invalid** if no type can be determined

**Type Resolution with Extensions:**
Since `$extends` follows JSON Schema `$ref` semantics, type inheritance behavior is determined by constraint validation rather than explicit merge rules. Local type constraints are evaluated alongside inherited constraints according to JSON Schema validation patterns.

<aside class="example">

```json
{
  "base": {
    "$type": "color",
    "primary": {
      "$value": {
        "colorSpace": "srgb",
        "components": [0, 0.4, 0.8],
        "hex": "#0066cc"
      }
    }
  },
  "extended": {
    "$extends": "{base}",
    "$type": "dimension", // Local constraint
    "spacing": { "$value": { "value": 16, "unit": "px" } }
  }
}
```

</aside>

In this example, the group `extended` must satisfy both its local `$type: "dimension"` constraint and any applicable constraints from the referenced `base` group, following JSON Schema constraint resolution rules.

### Circular Reference Detection

> Tools MUST detect and report circular references in:

Reviewer: Optional, ignorable: Should “report” be more explicit?

```suggestion
Tools MUST detect and throw an error on circular references in:
```
